In [1]:
import numpy as np
import xarray as xr
import pandas as pd

In [2]:
import datetime

In [3]:
da = xr.DataArray(
    np.random.randint(1000, size=1827),
    coords=[pd.date_range("2000-01-01", "2004-12-31", freq="D")],
    dims="time",)

In [4]:
da

<xarray.DataArray (time: 1827)>
array([792, 522, 369, ..., 578, 569, 175])
Coordinates:
  * time     (time) datetime64[ns] 2000-01-01 2000-01-02 ... 2004-12-31

PeakEeveyYear is the maximum value in that year.

In [5]:
PeakEveryYear = da.groupby("time.year").max()

In [6]:
PeakEveryYear

<xarray.DataArray (year: 5)>
array([999, 996, 987, 989, 999])
Coordinates:
  * year     (year) int64 2000 2001 2002 2003 2004

Here we test how to find the time with peak value in the group of 2000 year

In [7]:
(da.groupby("time.year")[2000].where(da.groupby("time.year")[2000]==PeakEveryYear[0], drop=True)).time

<xarray.DataArray 'time' (time: 1)>
array(['2000-12-30T00:00:00.000000000'], dtype='datetime64[ns]')
Coordinates:
  * time     (time) datetime64[ns] 2000-12-30
    year     int64 2000

So we see the date is 2000-12-30 with the peak value of 999. Then we can get the day of the year by 'time.dt.dayofyear'.

In [8]:
# ind1 is the day of the year of 2000. 
ind1=(da.groupby("time.year")[2000].where(da.groupby("time.year")[2000]==PeakEveryYear[0], drop=True)).time.dt.dayofyear.values
print(ind1)

[365]


In [9]:
len(da.groupby('time.year')) #how many groups

5

In [11]:
ii=0
ind=np.zeros(len(da.groupby('time.year')))
for label, group in da.groupby('time.year'):
    print(label)
    ind[ii]=da.groupby("time.year")[label].where(da.groupby("time.year")[label]==PeakEveryYear[ii], drop=True).time[0].dt.dayofyear.values
    ii=ii+1
    
print(ind)

2000
2001
2002
2003
2004
[365. 264.  16.  73.  65.]


This result means in 2000 the peak value is on 365 days of the year.
                  in 2001 the peak value is on 264 days of the year....